### PR TITLE
[SPARK-40533] [CONNECT] [PYTHON] Support most built-in literal types for Python in Spark Connect

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?

This PR implements the client-side serialization of most Python literals into Spark Connect literals. 

### Why are the changes needed?
Expanding the Python client support.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT

### CODE REVIEW

**Comment by grundprinzip** [comment]
R: @HyukjinKwon @zhengruifeng @amaliujia

File: python/pyspark/sql/connect/column.py [hunk 1]
```diff
@@ -99,11 +101,59 @@ def to_plan(self, session: Optional["RemoteSparkSession"]) -> "proto.Expression"
...
+        elif value_type is datetime.time:
+            # Nanoseconds of the day.
+            tv = cast(datetime.time, self._value)
+            offset = (tv.second + tv.minute * 60 + tv.hour * 3600) * 1000 + tv.microsecond
+            exp.literal.time = int(offset * 1000)
+        elif value_type is datetime.date:
+            # Days since epoch.
+            days_since_epoch = (cast(datetime.date, self._value) - datetime.date(1970, 1, 1)).days
+            exp.literal.date = days_since_epoch
+        elif value_type is uuid.UUID:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/column.py:134]: Hm, this isn't actually supported in current PySpark's `lit`. Should we maybe exclude this for now?

File: python/pyspark/sql/connect/_typing.py [hunk 2]
```diff
@@ -15,5 +15,7 @@
 # limitations under the License.
 #
 from typing import Union
+from datetime import date, time, datetime
 
 PrimitiveType = Union[str, int, bool, float]
+LiteralType = Union[PrimitiveType, Union[date, time, datetime]]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/_typing.py:21]: Seems not used.

File: python/pyspark/sql/connect/column.py (hunk 1 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/column.py:134]: ack, will remove.

Comment (grundprinzip) [reply python/pyspark/sql/connect/column.py:134]: Done.

**Comment by grundprinzip** [comment]
@HyukjinKwon removed the UUID support, please have another look.

File: python/pyspark/sql/connect/column.py [hunk 3]
```diff
@@ -99,11 +101,59 @@ def to_plan(self, session: Optional["RemoteSparkSession"]) -> "proto.Expression"
...
             exp.literal.string = cast(str, self._value)
         elif value_type is float:
             exp.literal.fp64 = cast(float, self._value)
+        elif value_type is decimal.Decimal:
+            d_v = cast(decimal.Decimal, self._value)
+            v_tuple = d_v.as_tuple()
+            exp.literal.decimal.scale = abs(v_tuple.exponent)
+            exp.literal.decimal.precision = len(v_tuple.digits) - abs(v_tuple.exponent)
+            # Two complement yeah...
+            raise ValueError("Python Decimal not supported.")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/column.py:117]: Can we remove if this is not implemented yet?

File: python/pyspark/sql/connect/column.py [hunk 4]
```diff
@@ -99,11 +101,59 @@ def to_plan(self, session: Optional["RemoteSparkSession"]) -> "proto.Expression"
...
+        elif value_type is datetime.time:
+            # Nanoseconds of the day.
+            tv = cast(datetime.time, self._value)
+            offset = (tv.second + tv.minute * 60 + tv.hour * 3600) * 1000 + tv.microsecond
+            exp.literal.time = int(offset * 1000)
+        elif value_type is datetime.date:
+            # Days since epoch.
+            days_since_epoch = (cast(datetime.date, self._value) - datetime.date(1970, 1, 1)).days
+            exp.literal.date = days_since_epoch
+        elif value_type is uuid.UUID:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/column.py:134]: Maybe we could remove `elif` so `else` branch throw the exception?

File: python/pyspark/sql/connect/column.py [hunk 5]
```diff
@@ -99,11 +101,59 @@ def to_plan(self, session: Optional["RemoteSparkSession"]) -> "proto.Expression"
...
+            exp.literal.decimal.precision = len(v_tuple.digits) - abs(v_tuple.exponent)
+            # Two complement yeah...
+            raise ValueError("Python Decimal not supported.")
+        elif value_type is bytes:
+            exp.literal.binary = self._value
+        elif value_type is datetime.datetime:
+            # Microseconds since epoch.
+            dt = cast(datetime.datetime, self._value)
+            v = dt - datetime.datetime(1970, 1, 1, 0, 0, 0, 0)
+            exp.literal.timestamp = int(v / datetime.timedelta(microseconds=1))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/column.py:124]: Can we maybe match the implementation in PySpark's: https://github.com/apache/spark/blob/master/python/pyspark/sql/types.py#L254-L260 ?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

Let's address complete types in a followup.

**Pull request closed without merging**